src/VisualStudio/Core/Def/LanguageService/AbstractPackage`2.cs

```diff
@@ -34,11 +34,18 @@ protected AbstractPackage()
     {
     }
 
-    protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
+    protected override void RegisterInitializationWork(PackageRegistrationTasks packageRegistrationTasks)
     {
-        await base.InitializeAsync(cancellationToken, progress).ConfigureAwait(false);
+        base.RegisterInitializationWork(packageRegistrationTasks);
 
-        await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+        packageRegistrationTasks.AddTask(isMainThreadTask: true, task: PackageInitializationMainThreadAsync);
+    }
+
+    private async Task PackageInitializationMainThreadAsync(IProgress<ServiceProgressData> progress, PackageRegistrationTasks packageRegistrationTasks, CancellationToken cancellationToken)
+    {
+        // This code uses various main thread only services, so it must run completely on the main thread
+        // (thus the CA(true) usage throughout)
```

Comment:
This comment to me is dubious to me at this point:

1. The solution local isn't being used, so we can delete that.
2. The RegisterLanguageService is just registering services so it's free-threaded now, right?
3. Creating the editor factories there is just creating some types...

The only things that potentially still are suspicious are registering the editor factories and that IsInCommandLineMode() line.

Reply:
1) -- Yep, that can easily be deleted
2) -- It looks like that can be moved to a bgthread registration
3) -- Creating the editor factories *might* be ok on a bg thread, but the registration needs to be on the main thread.

Will definitely clean up this are some more in future PRs.